File: generate_random_id.py
# python 3.13.2

import sys
import string
import random



def getCLIArgumentValue(key):
    i = sys.argv.index(key)

    if ((i + 1) >= len(sys.argv)):
        return None
    else:
        return sys.argv[i + 1]

def getRandomCharInString(s):
    i = random.randint(0, (len(s) - 1))
    return s[i]

def containsOneOfSubstrings(s, subs):
    for substring in subs:
        if substring in s:
            return True
    return False



args = sys.argv
defLen = "8"
defMode = "lLd"

if not (len(args) - 1):
    args_len = input("Length of id : ") or defLen
    args_mode = input("Mode type : ") or defMode

    args.extend([
        '-len', args_len,
        '-mode', args_mode,
    ])

if not args.count('-len'):
    args.extend([ '-len', defLen ])
if not args.count('-mode'):
    args.extend([ '-mode', defMode ])


# -info
if args.count('-info'):
    print("""
Build on python 3.13.2

Also, fun fact : when generating 8-symbols id using only lowercase letters the
probability of getting two exactly identical id's is ~4.79e-12 (or approx.
0.0000000005%).
    """)
    sys.exit()

# -help
if (args.count('-help') or args.count('-elp')):
    print("""
CLI options :
     -len [number]  Length of the id
     -mode [type]   Type of included symbols
     -info          Prints info about this program
     -help, -elp    Prints this message


The value of -mode [type] argument must be combination of flags as listed here
below. If omitted, takes default ('lLd').
    'l' = lower case letters
    'L' = upper case letters
    'd' = digits
    'h' = hexidigits
    'o' = octdigits

For example, flag 'lLh' will generate an id with both case letters and
hexidigits.

Or, value of -mode [type] argument can be 'custom=[symbols]', i.e. user-defined
combination of symbols.
    """)
    sys.exit()


# length
try:
    id_length = int( getCLIArgumentValue('-len') )
except Exception as err:
    print( str(err) )
    sys.exit()

# mode
s = ""
argValue = getCLIArgumentValue('-mode')

if argValue[0:7] == 'custom=':
    s = argValue[7:]
elif containsOneOfSubstrings(argValue, string.punctuation):
    print(f"Mode '{ argValue }' is not defined.")
    sys.exit()
else:
    if 'l' in argValue:
        s = "".join([s, string.ascii_lowercase])
    if 'L' in argValue:
        s = "".join([s, string.ascii_uppercase])
    if 'd' in argValue:
        s = "".join([s, string.digits])
    if 'h' in argValue:
        s = "".join([s, string.hexdigits])
    if 'o' in argValue:
        s = "".join([s, string.octdigits])

if not len(s):
    print(f"Mode '{ argValue }' is not defined.")
    sys.exit()


a = ""

for n in range(id_length):
    a += getRandomCharInString(s)

print(a)

File: generate_random_id.dart
import 'dart:math';



void main(List<String> argv) {

    if (argv.length == 0) return print('No CLI arguments provided (use \'-help\' for more information).\n');

    if (argv.contains('-info')) {

        print(
            'Build on Dart SDK v3.8.1\n\n'
            'Also, fun fact : when generating 8-symbols id using only lowercase letters\n'
            'the probability of getting two exactly identical id\'s is ~4.79e-12 (or\napprox. 0.0000000005%).\n'
        );

        return;

    }

    if (argv.contains('-help') || argv.contains('-elp')) {

        print(
            'CLI arguments :\n'
            '  -len [number] = length of the id\n'
            '  -mode [type] = type of included symbols\n'
            '  -info = print info about this program\n'
            '  -help / -elp = print this message\n\n'
            'The value of -mode [type] argument must be combination of flags as listed here\n'
            'below. If omitted, takes default (\'lLd\').\n'
            '  \'l\' = lower case letters\n'
            '  \'L\' = upper case letters\n'
            '  \'d\' = digits\n'
            '  \'h\' = hexidigits\n'
            '  \'o\' = octdigits\n'
            'for example :\n'
            '  flag \'lLh\' will generate an id with both case letters and hexidigits.\n'
        );

        return;

    }



    int len = 0;

    try {

        len = int.parse( getCLIArgumentValue(argv, 'len') );

    } catch (err) {

        print('-len parameter value is not a number (reading "${ getCLIArgumentValue(argv, 'len') }").\n');

        return;

    }


    String s = '';

    if (getCLIArgumentValue(argv, 'mode').contains('l')) s += 'abcdefghijklmnopqrstuvwxyz';
    if (getCLIArgumentValue(argv, 'mode').contains('L')) s += 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    if (getCLIArgumentValue(argv, 'mode').contains('d')) s += '0123456789';
    if (getCLIArgumentValue(argv, 'mode').contains('h')) s += '0123456789abcdefABCDEF';
    if (getCLIArgumentValue(argv, 'mode').contains('o')) s += '01234567';

    if (getCLIArgumentValue(argv, 'mode') == '') s += 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';

    if (s.length == 0) {

        print('Mode "${ getCLIArgumentValue(argv, 'mode') }" is not defined.\n');

        return;

    }


    String a = '';

    for (int i = 0; i < len; i++) {

        a += getIterableRandomPositionValue(s);

    }

    print(a);

}



String getCLIArgumentValue(List<String> args, String key) {

    if (args.indexOf('-$key') == -1) return '';

    return args[args.indexOf('-$key') + 1];

}

dynamic getIterableRandomPositionValue(iterable) {

    int i = Random().nextInt(iterable.length - 1);

    return iterable[i];

}
